### ISSUE TITLE
sshandle still have memory leak, tested with 4.0.24, 4.0.25 and 4.0.17.

### ISSUE THREAD

**Issue by devsunny** [netty/netty#3447]
java.lang.OutOfMemoryError: Direct buffer memory
	at java.nio.Bits.reserveMemory(Unknown Source)
	at java.nio.DirectByteBuffer.<init>(Unknown Source)
	at java.nio.ByteBuffer.allocateDirect(Unknown Source)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.allocateDirect(UnpooledUnsafeDirectByteBuf.java:108)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:69)
	at io.netty.buffer.UnpooledByteBufAllocator.newDirectBuffer(UnpooledByteBufAllocator.java:50)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:155)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:146)
	at io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:83)
	at io.netty.handler.ssl.SslHandler.allocate(SslHandler.java:1404)
	at io.netty.handler.ssl.SslHandler.allocateOutNetBuf(SslHandler.java:1414)
	at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:476)
	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:454)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:663)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:644)
	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:663)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:644)
	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:663)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:693)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:681)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:716)

**Comment by devsunny** [comment]
Sun JDK 1.7
OS: RHEL 6.4

**Comment by normanmaurer** [comment]
@devsunny we need more infos.. can you submit a reproducer ? Also maybe you just write to fast ?

**Comment by devsunny** [comment]
I retrieved about thirteen thousand records from JDBC and send through SSLHandler, then I got OOM. Digging through the code, I set io.netty.tryUnsafe to false to force it to use heap instead of direct memory, then it works fine. I will try to create a simple reproducer. I think you were right. It could be too fast; on the other hand, if direct memory buffer could not support fast write, what is the point to use direct memory buffer.

**Comment by Scottmitch** [comment]
@devsunny - Reproducer would be very helpful.  What are your JVM args, system memory details, and which Netty allocator are you using?  Are your 13k JDBC query responses in direct buffers?

**Comment by doom369** [comment]
4.0.26.Final
Just got same issue on my server. Server had small load ~30 req/sec.

Here is my stack :
```
java.lang.OutOfMemoryError: Direct buffer memory
        at java.nio.Bits.reserveMemory(Bits.java:658) ~[?:1.8.0_31]
        at java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:123) ~[?:1.8.0_31]
        at java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:311) ~[?:1.8.0_31]
        at io.netty.buffer.UnpooledUnsafeDirectByteBuf.allocateDirect(UnpooledUnsafeDirectByteBuf.java:108) ~[server-0.2.jar:?]
        at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:69) ~[server-0.2.jar:?]
        at io.netty.buffer.UnpooledByteBufAllocator.newDirectBuffer(UnpooledByteBufAllocator.java:50) ~[server-0.2.jar:?]
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:155) ~[server-0.2.jar:?]
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:146) ~[server-0.2.jar:?]
        at io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:83) ~[server-0.2.jar:?]
        at io.netty.handler.ssl.SslHandler.allocate(SslHandler.java:1468) ~[server-0.2.jar:?]
        at io.netty.handler.ssl.SslHandler.allocateOutNetBuf(SslHandler.java:1478) ~[server-0.2.jar:?]
        at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:491) ~[server-0.2.jar:?]
        at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:469) ~[server-0.2.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:735) [server-0.2.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:716) [server-0.2.jar:?]
        at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115) [server-0.2.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:735) [server-0.2.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.access$1500(AbstractChannelHandlerContext.java:32) [server-0.2.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.write(AbstractChannelHandlerContext.java:1033) [server-0.2.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:965) [server-0.2.jar:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:357) [server-0.2.jar:?]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357) [server-0.2.jar:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111) [server-0.2.jar:?]
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) [server-0.2.jar:?]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_31]

```

I'll try to create reproducer and will update issue.

**Comment by doom369** [comment]
A started performing load test and got same stacktrace on client side and wasn't able to reproduce on server. So your suggestion regarding "you just write to fast" could be true... 
But this means netty SSLHandler not able to handle sudden increase in load, even those sudden increase is few dozens of requests in my case.

**Comment by normanmaurer** [comment]
I think this was fixed in #3567. Please re-open if still present with 4.0.27.Final
